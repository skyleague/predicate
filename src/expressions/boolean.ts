import { inspect } from 'node:util'
import { equal } from '@skyleague/axioms'
import { type FactsFomExprs, operator } from '../engine/operator.js'
import {
    type AsExpression,
    type Expression,
    type ExpressionTypeOfLiteral,
    fromLiteral,
    type InputFromExpressions,
    type LiteralOr,
    type ValueExpression,
} from '../engine/types.js'
import { $value, type ValueItem } from './higher-order-fn.js'

export const $startsWith = operator({
    operator: 'startsWith',
    symbol: '$startsWith',
    fn: ([str, searchString]: [str: string, searchString: string]) => str.startsWith(searchString),
})
export const $endsWith = operator({
    operator: 'endsWith',
    symbol: '$endsWith',
    fn: ([str, searchString]: [str: string, searchString: string]) => str.endsWith(searchString),
})
export const $includes = operator({
    operator: 'includes',
    symbol: '$includes',
    fn: ([str, searchString]: [str: string, searchString: string]) => str.includes(searchString),
})

export const $equal = operator({ operator: '==', symbol: '$equal', fn: ([a, b]: [unknown, unknown]) => equal(a, b) })

export const $not = operator({ operator: '~', symbol: '$not', fn: ([x]: [boolean]) => !x })

export const $gt = operator({ operator: '>', symbol: '$gt', fn: ([a, b]: [number, number]) => a > b })
export const $gte = operator({ operator: '>=', symbol: '$gte', fn: ([a, b]: [number, number]) => a >= b })
export const $lt = operator({ operator: '<', symbol: '$lt', fn: ([a, b]: [number, number]) => a < b })
export const $lte = operator({ operator: '<=', symbol: '$lte', fn: ([a, b]: [number, number]) => a <= b })

export const $all = Object.assign(
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    <Expr extends LiteralOr<any[]>, Out extends Expression<boolean>>(
        xs: Expr,
        predicate: (value: ValueItem<ExpressionTypeOfLiteral<Expr>>) => Out,
    ): ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>> => {
        const _xs = fromLiteral(xs)
        const _value = $value<ExpressionTypeOfLiteral<Expr>>()
        const _transform = predicate(_value)
        return {
            fn: ([ys], ctx) => {
                return ys.every((x) => ctx.withScope(x, () => ctx.evaluate(_transform)))
            },
            dependsOn: [_xs],
            expr: (mode) => ({
                all: [_xs.expr(mode), _transform.expr(mode)],
            }),
            [inspect.custom]() {
                return `$all(${_xs[inspect.custom]?.() ?? ''}, (x) => ${_transform[inspect.custom]?.() ?? ''})`
            },
        } as ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>>
    },
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    { operator: 'all', symbol: '$all', parse: (xs: any, predicate: any) => $all(xs, () => predicate) } as const,
)

export const $any = Object.assign(
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    <Expr extends LiteralOr<any[]>, Out extends Expression<boolean>>(
        xs: Expr,
        predicate: (value: ValueItem<ExpressionTypeOfLiteral<Expr>>) => Out,
    ): ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>> => {
        const _xs = fromLiteral(xs)
        const _value = $value<ExpressionTypeOfLiteral<Expr>>()
        const _transform = predicate(_value)
        return {
            fn: ([ys], ctx) => {
                return ys.some((x) => ctx.withScope(x, () => ctx.evaluate(_transform)))
            },
            dependsOn: [_xs],
            expr: (mode) => ({
                any: [_xs.expr(mode), _transform.expr(mode)],
            }),
            [inspect.custom]() {
                return `$any(${_xs[inspect.custom]?.() ?? ''}, (x) => ${_transform[inspect.custom]?.() ?? ''})`
            },
        } as ValueExpression<boolean, InputFromExpressions<[AsExpression<Expr>]>, FactsFomExprs<Expr | Out>>
    },
    // biome-ignore lint/suspicious/noExplicitAny: this is needed for greedy matching
    { operator: 'any', symbol: '$any', parse: (xs: any, predicate: any) => $any(xs, () => predicate) } as const,
)
