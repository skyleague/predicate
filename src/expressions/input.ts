import type {
    FactExpression,
    DefinitionType,
    Expression,
    InputExpression,
    InferExpressionType,
    LiteralExpression,
} from '../engine/types.js'
import type { FromExpr } from '../json/jsonexpr.type.js'

import { type JSONPathValue, JSONPath } from '@skyleague/jsonpath'
import type { Schema } from '@skyleague/therefore'

import { inspect } from 'node:util'

export interface Fact<T, Name extends string> extends FactExpression<T> {
    name: Name
    dependsOn: []
    _type: 'fact'
}

export function $fact<T, Name extends string>(schema: Pick<Schema<T>, 'schema' | 'is'>, name: Name): Fact<T, Name> {
    return {
        _type: 'fact',
        dependsOn: [],
        name: name,
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => {
            return ctx.input[name]
        }) as Fact<T, Name>['fn'],
        expr: (definition: DefinitionType) =>
            (definition === 'definition' ? schema.schema : { fact: name.toString() }) as InferExpressionType<T>,
        [inspect.custom]() {
            return `$fact({schema: ${JSON.stringify(schema.schema)}}, "${name}")`
        },
    }
}
export interface From<I, O, DependsOn extends Expression[]> extends InputExpression<O, I, DependsOn, FromExpr> {
    _type: 'value'
}

export function $from<T, Name extends string>(fact: Fact<T, Name>): From<T, JSONPathValue<T, '$'>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P = '$' as P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]> {
    return {
        _type: 'value',
        dependsOn: [fact],
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => JSONPath.get(ctx.input[fact.name], path)) as From<
            T,
            JSONPathValue<T, P>,
            [Fact<T, Name>]
        >['fn'],
        expr: (_definition) => ({ from: [fact.name, path.toString()] as const }),
        [inspect.custom]() {
            return `$from(${fact[inspect.custom]?.() ?? `"${fact.name}"`}${path === '$' ? '' : `, "${path}"`})`
        },
    }
}

export function $literal<O>(x: O): LiteralExpression<O> {
    return {
        fn: () => x,
        dependsOn: [],
        expr: () => x as InferExpressionType<O>,
        _type: 'literal',
        [inspect.custom]() {
            return `$literal(${inspect(x)})`
        },
    }
}
