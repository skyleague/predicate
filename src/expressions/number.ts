import { operator } from '../engine/operator.js'

export const $add = operator({ operator: '+', symbol: '$add', fn: ([a, b]: [number, number]) => a + b })
export const $subtract = operator({ operator: '-', symbol: '$subtract', fn: ([a, b]: [number, number]) => a - b })
export const $divide = operator({ operator: '/', symbol: '$divide', fn: ([a, b]: [number, number]) => a / b })
export const $multiply = operator({ operator: '*', symbol: '$multiply', fn: ([a, b]: [number, number]) => a * b })
export const $modulo = operator({ operator: '%', symbol: '$modulo', fn: ([a, b]: [number, number]) => a % b })
