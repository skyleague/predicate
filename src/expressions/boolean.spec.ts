import { array, float, forAll, string, tuple } from '@skyleague/axioms'
import { arbitrary } from '@skyleague/therefore'
import { describe, expect, expectTypeOf, it } from 'vitest'
import { LogicObj } from '../../test/logic.type.js'
import { AbObj } from '../../test/string.type.js'
import { $policy } from '../engine/policy.js'
import type { ValueExpression } from '../engine/types.js'
import type { BooleanExpr, EndsWithExpr, IncludesExpr, StartsWithExpr } from '../json/jsonexpr.type.js'
import { $all, $any, $endsWith, $equal, $includes, $startsWith } from './boolean.js'
import { $fact, $from, $literal, type Fact } from './input.js'
import { $modulo } from './number.js'

describe('startsWith', () => {
    it('handles a simple schema case', () => {
        const input = {
            a: '1',
            b: '2',
        }
        const fact = $fact(AbObj, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $startsWith(x1, x2)
        const startsWith = $policy({ x1, x2, x3 })
        const result = startsWith.evaluate({ input })
        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": "1",
                "b": "2",
              },
            },
            "output": {
              "x1": "1",
              "x2": "2",
              "x3": false,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<AbObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: string; x2: string; x3: boolean }>()
        expect(startsWith.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(AbObj, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $startsWith(x1, x2)
        const startsWith = $policy({ x1, x2, x3 })
        forAll(arbitrary(AbObj), (input) => {
            const result = startsWith.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({ x1: input.a, x2: input.b, x3: result.output.x1.startsWith(result.output.x2) })
            expectTypeOf(result.input.input).toEqualTypeOf<AbObj>()
            expectTypeOf(result.output).toEqualTypeOf<{ x1: string; x2: string; x3: boolean }>()
        })
    })

    it('handles two strings', () => {
        const x1 = $startsWith('1', '2')
        const startsWith = $policy({ x1 })
        const result = startsWith.evaluate()
        expect(result).toMatchInlineSnapshot(`
          {
            "input": undefined,
            "output": {
              "x1": false,
            },
          }
        `)
        expectTypeOf(result.output).toEqualTypeOf<{ x1: boolean }>()
        expect(startsWith.expr()).toMatchSnapshot()
    })

    it('handles any two strings', () => {
        forAll(tuple(string({ minLength: 1 }), string({ minLength: 1 })), ([a, b]) => {
            const x1 = $startsWith(a + b, a)
            const startsWith = $policy({ x1 })
            const result = startsWith.evaluate()
            expect(result.output.x1).toEqual(true)
        })
    })

    it('handles the types', () => {
        const fact = $fact(AbObj, 'input')

        const x1 = $startsWith('1', '2')
        expectTypeOf(x1).toEqualTypeOf<ValueExpression<boolean, [str: '1', searchString: '2'], never, StartsWithExpr>>()

        const x2 = $startsWith($from(fact, '$.a'), '2')
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<boolean, [str: string, searchString: '2'], [Fact<AbObj, 'input'>], StartsWithExpr>
        >()

        const x3 = $startsWith('2', $from(fact, '$.b'))
        expectTypeOf(x3).toEqualTypeOf<
            ValueExpression<boolean, [str: '2', searchString: string], [Fact<AbObj, 'input'>], StartsWithExpr>
        >()
    })
})

describe('endsWith', () => {
    it('handles a simple schema case', () => {
        const input = {
            a: '1',
            b: '2',
        }
        const fact = $fact(AbObj, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $endsWith(x1, x2)
        const endsWith = $policy({ x1, x2, x3 })
        const result = endsWith.evaluate({ input })
        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": "1",
                "b": "2",
              },
            },
            "output": {
              "x1": "1",
              "x2": "2",
              "x3": false,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<AbObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: string; x2: string; x3: boolean }>()
        expect(endsWith.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(AbObj, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $endsWith(x1, x2)
        const startsWith = $policy({ x1, x2, x3 })
        forAll(arbitrary(AbObj), (input) => {
            const result = startsWith.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({ x1: input.a, x2: input.b, x3: result.output.x1.endsWith(result.output.x2) })
            expectTypeOf(result.input.input).toEqualTypeOf<AbObj>()
            expectTypeOf(result.output).toEqualTypeOf<{ x1: string; x2: string; x3: boolean }>()
        })
    })

    it('handles two strings', () => {
        const x1 = $endsWith('1', '2')
        const startsWith = $policy({ x1 })
        const result = startsWith.evaluate()
        expect(result).toMatchInlineSnapshot(`
          {
            "input": undefined,
            "output": {
              "x1": false,
            },
          }
        `)
        expectTypeOf(result.output).toEqualTypeOf<{ x1: boolean }>()
        expect(startsWith.expr()).toMatchSnapshot()
    })

    it('handles any two strings', () => {
        forAll(tuple(string({ minLength: 1 }), string({ minLength: 1 })), ([a, b]) => {
            const x1 = $endsWith(a + b, b)
            const startsWith = $policy({ x1 })
            const result = startsWith.evaluate()
            expect(result.output.x1).toEqual(true)
        })
    })

    it('handles the types', () => {
        const fact = $fact(AbObj, 'input')

        const x1 = $endsWith('1', '2')
        expectTypeOf(x1).toEqualTypeOf<ValueExpression<boolean, [str: '1', searchString: '2'], never, EndsWithExpr>>()

        const x2 = $endsWith($from(fact, '$.a'), '2')
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<boolean, [str: string, searchString: '2'], [Fact<AbObj, 'input'>], EndsWithExpr>
        >()

        const x3 = $endsWith('2', $from(fact, '$.b'))
        expectTypeOf(x3).toEqualTypeOf<
            ValueExpression<boolean, [str: '2', searchString: string], [Fact<AbObj, 'input'>], EndsWithExpr>
        >()
    })
})

describe('includes', () => {
    it('handles a simple schema case', () => {
        const input = {
            a: '1',
            b: '2',
        }
        const fact = $fact(AbObj, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $includes(x1, x2)
        const endsWith = $policy({ x1, x2, x3 })
        const result = endsWith.evaluate({ input })
        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": "1",
                "b": "2",
              },
            },
            "output": {
              "x1": "1",
              "x2": "2",
              "x3": false,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<AbObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: string; x2: string; x3: boolean }>()
        expect(endsWith.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(AbObj, 'input')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $includes(x1, x2)
        const startsWith = $policy({ x1, x2, x3 })
        forAll(arbitrary(AbObj), (input) => {
            const result = startsWith.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({ x1: input.a, x2: input.b, x3: result.output.x1.includes(result.output.x2) })
            expectTypeOf(result.input.input).toEqualTypeOf<AbObj>()
            expectTypeOf(result.output).toEqualTypeOf<{ x1: string; x2: string; x3: boolean }>()
        })
    })

    it('handles two strings', () => {
        const x1 = $includes('1', '2')
        const startsWith = $policy({ x1 })
        const result = startsWith.evaluate()
        expect(result).toMatchInlineSnapshot(`
          {
            "input": undefined,
            "output": {
              "x1": false,
            },
          }
        `)
        expectTypeOf(result.output).toEqualTypeOf<{ x1: boolean }>()
        expect(startsWith.expr()).toMatchSnapshot()
    })

    it('handles any two strings', () => {
        forAll(tuple(string({ minLength: 1 }), string({ minLength: 1 })), ([a, b]) => {
            const x1 = $includes(a + b, b)
            const startsWith = $policy({ x1 })
            const result = startsWith.evaluate()
            expect(result.output.x1).toEqual(true)
        })
    })

    it('handles the types', () => {
        const fact = $fact(AbObj, 'input')

        const x1 = $includes('1', '2')
        expectTypeOf(x1).toEqualTypeOf<ValueExpression<boolean, [str: '1', searchString: '2'], never, IncludesExpr>>()

        const x2 = $includes($from(fact, '$.a'), '2')
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<boolean, [str: string, searchString: '2'], [Fact<AbObj, 'input'>], IncludesExpr>
        >()

        const x3 = $includes('2', $from(fact, '$.b'))
        expectTypeOf(x3).toEqualTypeOf<
            ValueExpression<boolean, [str: '2', searchString: string], [Fact<AbObj, 'input'>], IncludesExpr>
        >()
    })
})

describe('all', () => {
    it('handles a simple schema case', () => {
        const input = {
            c: true,
            a: 1,
            b: 2,
            d: [
                { a: true, b: false },
                { a: false, b: true },
            ],
            e: true,
        }
        const fact = $fact(LogicObj, 'input')
        const x1 = $from(fact, '$.d[*].a')
        const x2 = $from(fact, '$.d[*].b')
        const x3 = $all(x1, (x) => x)
        const hof = $policy({ x1, x2, x3 })
        const result = hof.evaluate({ input })

        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": 1,
                "b": 2,
                "c": true,
                "d": [
                  {
                    "a": true,
                    "b": false,
                  },
                  {
                    "a": false,
                    "b": true,
                  },
                ],
                "e": true,
              },
            },
            "output": {
              "x1": [
                true,
                false,
              ],
              "x2": [
                false,
                true,
              ],
              "x3": false,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: boolean[]; x2: boolean[]; x3: boolean }>()
        expect(hof.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(LogicObj, 'input')
        const x1 = $from(fact, '$.d[*].a')
        const x2 = $from(fact, '$.d[*].b')
        const d = $from(fact, '$.d')
        const x3 = $all(x1, (x) => $equal(true, x))
        const x4 = $all(x2, (x) => $equal(false, x))
        const x5 = $all(d, (x) => x('$.b'))
        const hof = $policy({ x1, x2, x3, x4, x5 })
        forAll(arbitrary(LogicObj), (input) => {
            const result = hof.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({
                x1: input.d.map((a) => a.a),
                x2: input.d.map((a) => a.b),
                x3: input.d.every((a) => a.a),
                x4: input.d.every((a) => !a.b),
                x5: input.d.every((a) => a.b),
            })
            expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
            expectTypeOf(result.output).toEqualTypeOf<{
                x1: boolean[]
                x2: boolean[]
                x3: boolean
                x4: boolean
                x5: boolean
            }>()
        })
    })

    it('handles any booleans', () => {
        forAll(array(float()), (xs) => {
            const x1 = $all(xs, (x) => $equal($modulo(x, 2), 1))
            const hof = $policy({ x1 })
            const result = hof.evaluate()
            expect(result.output.x1).toEqual(xs.every((x) => x % 2 === 1))
        })
    })

    it('handles the types', () => {
        const fact = $fact(LogicObj, 'input')

        const x1 = $all([1, 2], (x) => $equal(x, 1))
        expectTypeOf(x1).toEqualTypeOf<ValueExpression<boolean, [number[]], never, BooleanExpr>>()

        const a = $from(fact, '$.d')
        const x2 = $all(a, (x) => $equal(x, 1))
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<
                boolean,
                [
                    {
                        a: boolean
                        b: boolean
                    }[],
                ],
                [Fact<LogicObj, 'input'>],
                BooleanExpr
            >
        >()

        const da = $from(fact, '$.d..a')
        const x3 = $all(da, (x) => $equal(x, 1))
        expectTypeOf(x3).toEqualTypeOf<ValueExpression<boolean, [boolean[]], [Fact<LogicObj, 'input'>], BooleanExpr>>()
    })
})

describe('any', () => {
    it('handles a simple schema case', () => {
        const input = {
            c: true,
            a: 1,
            b: 2,
            d: [
                { a: true, b: false },
                { a: false, b: true },
            ],
            e: true,
        }
        const fact = $fact(LogicObj, 'input')
        const x1 = $from(fact, '$.d[*].a')
        const x2 = $from(fact, '$.d[*].b')
        const x3 = $any(x1, (x) => x)
        const hof = $policy({ x1, x2, x3 })
        const result = hof.evaluate({ input })

        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": 1,
                "b": 2,
                "c": true,
                "d": [
                  {
                    "a": true,
                    "b": false,
                  },
                  {
                    "a": false,
                    "b": true,
                  },
                ],
                "e": true,
              },
            },
            "output": {
              "x1": [
                true,
                false,
              ],
              "x2": [
                false,
                true,
              ],
              "x3": true,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: boolean[]; x2: boolean[]; x3: boolean }>()
        expect(hof.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(LogicObj, 'input')
        const x1 = $from(fact, '$.d[*].a')
        const x2 = $from(fact, '$.d[*].b')
        const d = $from(fact, '$.d')
        const x3 = $any(x1, (x) => $equal(true, x))
        const x4 = $any(x2, (x) => $equal(false, x))
        const x5 = $any(d, (x) => x('$.b'))
        const hof = $policy({ x1, x2, x3, x4, x5 })
        forAll(arbitrary(LogicObj), (input) => {
            const result = hof.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({
                x1: input.d.map((a) => a.a),
                x2: input.d.map((a) => a.b),
                x3: input.d.some((a) => a.a),
                x4: input.d.some((a) => !a.b),
                x5: input.d.some((a) => a.b),
            })
            expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
            expectTypeOf(result.output).toEqualTypeOf<{
                x1: boolean[]
                x2: boolean[]
                x3: boolean
                x4: boolean
                x5: boolean
            }>()
        })
    })

    it('handles any booleans', () => {
        forAll(array(float()), (xs) => {
            const x1 = $any(xs, (x) => $equal($modulo(x, 2), 1))
            const hof = $policy({ x1 })
            const result = hof.evaluate()
            expect(result.output.x1).toEqual(xs.some((x) => x % 2 === 1))
        })
    })

    it('handles the types', () => {
        const fact = $fact(LogicObj, 'input')

        const x1 = $any([1, 2], (x) => $equal(x, 1))
        expectTypeOf(x1).toEqualTypeOf<ValueExpression<boolean, [number[]], never, BooleanExpr>>()

        const a = $from(fact, '$.d')
        const x2 = $any(a, (x) => $equal(x, 1))
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<
                boolean,
                [
                    {
                        a: boolean
                        b: boolean
                    }[],
                ],
                [Fact<LogicObj, 'input'>],
                BooleanExpr
            >
        >()

        const da = $from(fact, '$.d..a')
        const x3 = $any(da, (x) => $equal(x, 1))
        expectTypeOf(x3).toEqualTypeOf<ValueExpression<boolean, [boolean[]], [Fact<LogicObj, 'input'>], BooleanExpr>>()

        const x4 = $any($literal(['foo']), (x) => $equal(x, a))
        expectTypeOf(x4).toEqualTypeOf<ValueExpression<boolean, [string[]], [Fact<LogicObj, 'input'>], BooleanExpr>>()
    })
})
