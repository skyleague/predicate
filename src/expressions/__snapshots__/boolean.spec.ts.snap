// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`all > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "b": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "c": {
          "type": "boolean",
        },
        "d": {
          "items": {
            "additionalProperties": true,
            "properties": {
              "a": {
                "type": "boolean",
              },
              "b": {
                "type": "boolean",
              },
            },
            "required": [
              "a",
              "b",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "e": {
          "type": "boolean",
        },
      },
      "required": [
        "c",
        "a",
        "b",
        "d",
        "e",
      ],
      "title": "LogicObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.d[*].a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.d[*].b",
      ],
    },
    "x3": {
      "all": [
        {
          "from": [
            "input",
            "$.d[*].a",
          ],
        },
        {
          "value": "$",
        },
      ],
    },
  },
}
`;

exports[`any > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "b": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "c": {
          "type": "boolean",
        },
        "d": {
          "items": {
            "additionalProperties": true,
            "properties": {
              "a": {
                "type": "boolean",
              },
              "b": {
                "type": "boolean",
              },
            },
            "required": [
              "a",
              "b",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "e": {
          "type": "boolean",
        },
      },
      "required": [
        "c",
        "a",
        "b",
        "d",
        "e",
      ],
      "title": "LogicObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.d[*].a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.d[*].b",
      ],
    },
    "x3": {
      "any": [
        {
          "from": [
            "input",
            "$.d[*].a",
          ],
        },
        {
          "value": "$",
        },
      ],
    },
  },
}
`;

exports[`endsWith > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "type": "string",
        },
        "b": {
          "type": "string",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "title": "AbObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.b",
      ],
    },
    "x3": {
      "endsWith": [
        {
          "from": [
            "input",
            "$.a",
          ],
        },
        {
          "from": [
            "input",
            "$.b",
          ],
        },
      ],
    },
  },
}
`;

exports[`endsWith > handles two strings 2`] = `
{
  "input": {},
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "endsWith": [
        "1",
        "2",
      ],
    },
  },
}
`;

exports[`includes > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "type": "string",
        },
        "b": {
          "type": "string",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "title": "AbObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.b",
      ],
    },
    "x3": {
      "includes": [
        {
          "from": [
            "input",
            "$.a",
          ],
        },
        {
          "from": [
            "input",
            "$.b",
          ],
        },
      ],
    },
  },
}
`;

exports[`includes > handles two strings 2`] = `
{
  "input": {},
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "includes": [
        "1",
        "2",
      ],
    },
  },
}
`;

exports[`startsWith > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "type": "string",
        },
        "b": {
          "type": "string",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "title": "AbObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.b",
      ],
    },
    "x3": {
      "startsWith": [
        {
          "from": [
            "input",
            "$.a",
          ],
        },
        {
          "from": [
            "input",
            "$.b",
          ],
        },
      ],
    },
  },
}
`;

exports[`startsWith > handles two strings 2`] = `
{
  "input": {},
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "startsWith": [
        "1",
        "2",
      ],
    },
  },
}
`;
